### PR TITLE
Code refactor and minor optimization

### PR DESCRIPTION
**Pull request by GermanAizek**
@Olga-Yakovleva @maniyax @alex19EP ,
Minor correction and rewriting of code parts to more secure and correct according to the C++ standard, some minor code optimizations.

### CODE REVIEW

**Comment by GermanAizek** [comment]
```sh
[2021-10-04 09:13:41] [build-stdout] [ 29%] Linking CXX shared library libmage.so
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stdout] cd /opt/src/_lgtm_build_dir/src/third-party/mage && /usr/bin/cmake -E cmake_link_script CMakeFiles/libmage.dir/link.txt --verbose=1
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stdout] /usr/bin/c++ -fPIC -fpermissive -O3 -DNDEBUG -flto -fno-fat-lto-objects -Wl,-O1 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,-z,ibtplt -Wl,-z,ibt -Wl,-z,shstk -Wl,-z,notext -shared -Wl,-soname,libmage.so.2 -o libmage.so.2.0-RHVoice CMakeFiles/libmage.dir/FrameQueue.cpp.o CMakeFiles/libmage.dir/Label.cpp.o CMakeFiles/libmage.dir/LabelQueue.cpp.o CMakeFiles/libmage.dir/MathFunctions.cpp.o CMakeFiles/libmage.dir/Model.cpp.o CMakeFiles/libmage.dir/ModelQueue.cpp.o CMakeFiles/libmage.dir/Vocoder.cpp.o CMakeFiles/libmage.dir/fft-sptk.cpp.o CMakeFiles/libmage.dir/getmem-sptk.cpp.o CMakeFiles/libmage.dir/hts.cpp.o CMakeFiles/libmage.dir/mage.cpp.o CMakeFiles/libmage.dir/mlsacheck-sptk.cpp.o -Wl,-rpath,/opt/src/_lgtm_build_dir/src/third-party/hts_engine: ../hts_engine/libhts_engine.so.1.06-RHVoice -lRHVoice_question_match -lm 
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stderr] /usr/bin/ld: cannot find -lRHVoice_question_match
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stderr] collect2: error: ld returned 1 exit status
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stderr] make[2]: *** [src/third-party/mage/CMakeFiles/libmage.dir/build.make:253: src/third-party/mage/libmage.so.2.0-RHVoice] Error 1
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stdout] make[2]: Leaving directory '/opt/src/_lgtm_build_dir'
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stderr] make[1]: *** [CMakeFiles/Makefile2:154: src/third-party/mage/CMakeFiles/libmage.dir/all] Error 2
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stderr] make: *** [Makefile:155: all] Error 2
[2021-10-04 09:13:41] [build-stdout] [2021-10-04 09:13:41] [build-stdout] make[1]: Leaving directory '/opt/src/_lgtm_build_dir'
```

LGTM build failed

**Comment by alex19EP** [comment]
please. add a description of your changes.

File: src/lib/lib.cpp [hunk 1]
```diff
@@ -166,7 +166,7 @@ RHVoice_tts_engine_struct::RHVoice_tts_engine_struct(const RHVoice_init_params*
       const char** p=init_params->resource_paths;
       while(*p)
         {
-          engine_params.resource_paths.push_back(std::string(*p));
+          engine_params.resource_paths.emplace_back(*p);
```
Comment (alex19EP) [inline src/lib/lib.cpp:169]: not sure if I like this one for [code readability](https://abseil.io/tips/112)
reasons

**Comment by GermanAizek** [comment]
> 
> 
> please. add a description of your changes.
 up

**Comment by GermanAizek** [comment]
> 
> 
> please. add a description of your changes.

up

**Review by alex19EP — CHANGES_REQUESTED** [review]

File: src/lib/lib.cpp (hunk 1 above)
Comment (GermanAizek) [reply src/lib/lib.cpp:169]: @alex19EP,
U have a pointer declared in the constructor `std::string`, using the employment method you can reduce the size of the code, and there is not much to optimize.
U attached a link to the material that described the principle of the method very well, it is simply convenient and helps the compiler to choose the constructor move and copy for your object.

```cpp
void f2(const std::vector<std::string>& users) {
    std::vector<std::pair<std::string, int>> userOccurancies(users.size());

    auto userOcIt = userOccurancies.begin();
    for (const auto & user : users) {
        userOcIt->first = std::move(user);
        userOcIt->second = 0;
        userOcIt++;
    }
}

void f(const std::vector<std::string>& users) {
    std::vector < std::pair<std::string, std::size_t>> userCount;
    userCount.reserve(users.size());

    for (auto& user : users) {
        userCount.emplace_back(user, 0);
    }
}
```

above wrote an example code to better understand f2() function identical to f()

**Review by alex19EP — APPROVED** [review]

**Comment by alex19EP** [comment]
thank you for contrebution!

**Comment by Olga-Yakovleva** [comment]
Hi @GermanAizek. Just following up on the question of licensing your contributions.  Could you please let me know if you will license this contribution under a permissive license, or grant me a C.L.A?

[Here](https://github.com/RHVoice/contrib-licensing) is more about this topic.

**Comment by Olga-Yakovleva** [comment]
Unfortunately, there has been no answer to my question. I will have to replace this contribution with an alternative implementation of the same functionality. As this is more of a stylistic change, I may need to roll some things back for now.

**Pull request merged**